libcxx/test/std/experimental/reflection/reflect-invoke.pass.cpp

```diff
@@ -241,4 +239,59 @@ static_assert(
                          std::views::transform(std::meta::reflect_value<int>)));
 }  // namespace with_non_contiguous_ranges
 
+namespace non_static_member_functions {
+
+struct Number {
+public:
+  constexpr Number(int v) : value(v) {}
+
+  constexpr int plus(int a) const { return plus_impl(a); }
+
+  constexpr int get_value() const { return value; }
+
+  constexpr Number operator+(int num) const { return Number(plus_impl(num)); }
+
+  template <typename T>
+  constexpr T multiply(T x) const {
+    return value * x;
+  }
+
+  const int value;
+
+private:
+  constexpr int plus_impl(int a) const { return value + a; }
+};
+
+constexpr Number num{42};
+constexpr auto num_ref = &num;
+
+// member function with input arguments
+static_assert(std::meta::reflect_value(84) ==
+              reflect_invoke(^Number::plus,
+                             {^num, std::meta::reflect_value(42)}));
+
+// operator overload
+static_assert(std::meta::reflect_value(84) ==
+              reflect_invoke(^Number::get_value,
+                             {reflect_invoke(^Number::operator+,
+                                                {^num, std::meta::reflect_value(42)})}));
+
+// member function without input arguments
+static_assert(std::meta::reflect_value(42) ==
+              reflect_invoke(^Number::get_value,
+                             {^num}));
+
+// member function called with object reference
+static_assert(std::meta::reflect_value(42) ==
+              reflect_invoke(^Number::get_value,
+                             {^num_ref}));
+
+// template member function
+static_assert(std::meta::reflect_value(84) ==
+              reflect_invoke(^Number::multiply,
+                             {^int},
+                             {^num, std::meta::reflect_value(2)}));
+
+} // namespace non_static_member_functions
```

Comment:
Great tests! A few other suggestions:
1. Invoking `Base::fn()` with an object of type `Child`.
2. Passing a reflection of a pointer value pointing to a constexpr function.
3. Passing a reflection of an object with static storage duration holding a pointer to a constexpr function.

I'm actually not sure we have those tests for the existing `reflect_invoke` infrastructure either..

Reply:
Added tests for all listed cases + added few tests for pointers to non-member function.
However:
1. Test like this fails, which I think is not expected according to your test proposal 3 (if I understand task correctly, lol):
```cpp
constexpr int foo(int a) {
  return a + 42;
}

static int (*foo_static_pointer)(int) = &foo;
static_assert(reflect_invoke(^^foo_static_pointer, {std::meta::reflect_value(2)})
              == std::meta::reflect_value(44));
```
I could fix it in a separate PR if you agree that it's a valid bug
2. To work with pointers to member function, I added a small hack which tries to get `CXXMethodDecl *` from pointer. I did it for several reasons:

-  To reuse the same code path for validation.
-  To make it work with splices. If I pass just original `DeclRefExpr* DRE` to use it inside splice, I receive an error `error: use of splice as qualified member access requires the reflection to be a member of a class`, but with my hack everything works as expected. Maybe I haven't dig into code enough to understand the rootcase  and my method is a bad idea, opened for disscussion.

Reply:
That example should work if `foo_static_pointer` is constexpr, but not as written (otherwise _`INVOKE-EXPR`_ is not a constant expression). Does that explain the error you get for that test-case?

Reply:
My miss, fixed + covered with test.

However it still doesn't fix point 2 with error `error: use of splice as qualified member access requires the reflection to be a member of a class` so my hack is still in place but with small bugfix